### ISSUE TITLE
Add reconstruct_patches op + ReconstructPatches{2,3}D layers (inverse of extract_patches)

### ISSUE THREAD

**Issue by brianmanderson** [keras-team/keras#22951]
## Summary

Add the missing inverse of `keras.ops.image.extract_patches` to core Keras, in both op form
(`keras.ops.image.reconstruct_patches`, `keras.ops.image.reconstruct_patches_3d`) and Layer form
(`keras.layers.ReconstructPatches2D`, `keras.layers.ReconstructPatches3D`).

This PR is intentionally scoped to the basic non-overlapping case (`strides == size`, `channels_last`,
`padding="valid"`/`"same"`). Overlap, `channels_first`, `dilation_rate`, `reduction` modes, and other
features are planned as incremental follow-up PRs.

## Motivation

`extract_patches` (and `extract_patches_3d`) ship in Keras as forward-only ops. PyTorch has
`torch.nn.Unfold` and `torch.nn.Fold` as symmetric pair — Keras has the unfold half but no fold half.
This asymmetry comes up repeatedly in:

- **Patch-based autoencoders** that compress in patch space and need to reconstruct the original.
- **ViT decoders** that produce patch tokens which must be folded back to a pixel grid.
- **Sliding-window inference** where per-patch predictions need to be stitched into a full-resolution output.
- **3D medical-imaging** segmentation pipelines that operate on patches due to GPU-memory constraints
  and reassemble the full volume for downstream processing.

Reference: this addresses the gap that motivated #20046 (fold/unfold), which was closed without an
inverse op landing.

## Proposed API

```python
keras.ops.image.reconstruct_patches(
    patches,            # shape from extract_patches: (B, gH, gW, pH*pW*C) for 2D
    size,               # (pH, pW) — matches extract_patches' `size`
    output_size,        # (H, W) — original spatial shape before extraction
    strides=None,       # defaults to `size` (non-overlapping)
    padding="valid",    # "valid" | "same", matches extract_patches semantics
    data_format=None,
)

keras.ops.image.reconstruct_patches_3d(patches, size, output_size, ...)

keras.layers.ReconstructPatches2D(size, output_size, padding=..., data_format=..., name=...)
keras.layers.ReconstructPatches3D(size, output_size, padding=..., data_format=..., name=...)
```

Naming and argument names match the forward op exactly so the two compose cleanly. `output_size` is
required (not inferred) because under `padding="same"` the forward op is lossy in spatial dimension
(input `D=25` and `D=32` with patch 16 both produce `gD=2`), so the original `D` must be provided
explicitly. (A future PR can add `output_size=None` auto-infer for `padding="valid"`.)

## Implementation approach

Pure reshape / transpose / slice — no new ops, no math beyond layout. Specifically:

1. Reshape patches `(B, gH, gW, pH*pW*C)` → `(B, gH, gW, pH, pW, C)`
2. Transpose to `(B, gH, pH, gW, pW, C)`
3. Reshape to `(B, gH*pH, gW*pW, C)` — the padded canvas the forward conv would have seen
4. For `padding="same"`: slice off the padding using the same `pad_before = total // 2` convention
   that `backend.nn.conv("same")` uses on the forward pass (verified consistent across TF/JAX/torch).
5. For `padding="valid"`: no slicing.

This is the inverse of the conv-with-identity-kernel that `_extract_patches_*` does internally, but
implemented without `conv_transpose` for the non-overlap case — much faster (no math, just data
movement). Overlapping reconstruction (deferred to a follow-up PR) needs `conv_transpose` and is a
fundamentally different code path.

## Working branch

**Branch:** https://github.com/brianmanderson/keras/tree/pr/01-basic-reconstruct
**Diff:** 11 files changed, +851 lines (no deletions)

- `keras/src/ops/image.py` — adds `ReconstructPatches` Operation, `reconstruct_patches`,
  `reconstruct_patches_3d`, internal helpers
- `keras/src/layers/reshaping/reconstruct_patches{2,3}d.py` — Layer wrappers (new files)
- `keras/src/layers/reshaping/reconstruct_patches{2,3}d_test.py` — Layer tests (new files)
- `keras/src/ops/image_test.py` — symbolic-shape tests added next to the existing
  `test_extract_patches{,_3d}`
- `keras/src/layers/__init__.py` and the two `keras/api/.../layers/__init__.py` files — registrations
- `keras/api/{,_tf_keras/}ops/image/__init__.py` — op registrations

## Test status

All tests pass on TensorFlow, JAX, and PyTorch backends locally:

- 11 layer tests on `ReconstructPatches2D` covering valid + same padding, asymmetric patches,
  non-divisible inputs, get_config round-trip, error paths
- 11 layer tests on `ReconstructPatches3D` (same shape for 3D)
- 4 symbolic-shape tests on the ops in `image_test.py`

The killer test is `test_extract_then_reconstruct_roundtrip`, for several shape combinations
(including non-divisible ones forcing the `padding="same"` crop path), it asserts
`reconstruct(extract(x)) == x` exactly. This is the strongest argument the layer should exist; the
crop math is easy to get wrong by hand under `padding="same"` and our implementation matches the
forward op's padding distribution exactly across all three backends.

## Likely objection and pre-emption

The most common review pushback for "missing inverse" PRs is: *"users can compose this from
`keras.ops.reshape` + `transpose` + `slice` themselves."* Three counters:

1. **`padding="same"` is non-trivial.** The cropping must exactly match how `backend.nn.conv("same")`
   distributed padding on the forward pass. Getting this wrong silently shifts pixels by 1 — the
   roundtrip test catches it; manual code rarely does.
2. **Asymmetric APIs are a usability gap.** Shipping the forward op without an inverse forces every
   user to re-derive the same handful of lines, often incorrectly.
3. **PyTorch precedent.** `torch.nn.Fold` and `torch.nn.Unfold` are a symmetric pair, both first-class.
   This proposal brings Keras to parity.

## Follow-up PRs planned

If this lands, the planned incremental follow-ups (in dependency order) are:

1. Overlap support + `reduction="mean"`/`"sum"` (PyTorch `Fold` parity for overlapping patches)
2. `channels_first` data format
3. `dilation_rate` parameter
4. Auto-infer `output_size` for `padding="valid"` + improved error messages
5. `ExtractPatches2D`/`ExtractPatches3D` Layer wrappers
6. Dual-input dynamic `output_size` mode (for variable-input models)

Each of these has working code in a separate branch already. Mentioning here so reviewers can see
the trajectory without conflating scope into this PR.

**Comment by brianmanderson** [comment]
Thank you for the consideration!

**Comment by rehan243** [comment]
Oh nice, yeah, this would be super handy. Fwiw, we hit this gap when building a ViT-based pipeline for some image-to-image tasks. Reconstructing the original grid from patch embeddings was such a pain. We ended up rolling our own `reconstruct_patches` function using `tf.scatter_nd`. It works okay, but honestly, it felt hacky. 

Here’s a simplified version of what we did:  

```python
import tensorflow as tf

def reconstruct_patches(patches, grid_shape, patch_size):
    batch_size, grid_h, grid_w, patch_h, patch_w, channels = tf.shape(patches)
    target_shape = (batch_size, grid_h * patch_h, grid_w * patch_w, channels)
    
    # Reshape and scatter the patches
    patches_reshaped = tf.reshape(
        patches, [batch_size, grid_h, grid_w, patch_h, patch_w, channels]
    )
    grid = tf.scatter_nd(
        indices=tf.range(tf.reduce_prod(target_shape)),
        updates=tf.reshape(patches_reshaped, [-1]),
        shape=target_shape,
    )
    return grid
```

Super basic, doesn’t support overlapping patches and assumes `strides == size` (like you mentioned). But yeah, it’s wild that Keras doesn’t have this out of the box when PyTorch has `torch.nn.Fold`. Would love to see this land — stitching patches for sliding-window inference in 3D (medical imaging especially) is such a common workflow.

**Comment by brianmanderson** [comment]
Thanks for the interest! I'd made some custom layers a while back to try and address this gap but never felt confident enough to try and push the support up to keras itself. First time trying to be a contributor 😅

**Comment by rehan243** [comment]
Oh cool, so you've been working on custom layers! That’s awesome. Honestly, contributing to Keras can feel a bit daunting at first, but it's super rewarding once you get the hang of it. 

If you need a nudge, the Keras contribution guidelines are pretty helpful for getting started. And hey, if you've got a prototype for your layers, I'd love to see how it compares — maybe we can collaborate on it! What kind of issues did you run into when trying to push your changes?

**Comment by brianmanderson** [comment]
Thanks! Yeah the guideline was very helpful, without it I definitely would have done the faux-pas of simply making a pull request. The biggest issue was writing in correct tests to ensure functionality on edge cases.

I'm not really certain at this point when I'm supposed to provide the presented code as a pull request, from the guidelines it seems like it's okay to do now. Feel free to take a look at what I'm providing @rehan243!

**Comment by brianmanderson** [comment]
@dhantule hope you're doing well! I've covered the PR issues raised, please let me know if anything else is needed on my end or follow-up for needed corrections

**Comment by brianmanderson** [comment]
Hello, hope you're doing well. Reaching back out to see if there's anything wanted on my end. Thank you!

**Comment by brianmanderson** [comment]
@hertschuh @dhantule Hope you're both doing well. I've updated the PR ticket to explain why I included both the layer option and ops. Happy to trim if wanted. Just realized that a few weeks have passed and I don't want this to drop off the board
Thank you!

**Comment by brianmanderson** [comment]
`keras.ops.image.reconstruct_patches` merged in #22984 (thanks @hertschuh!). Scope note: per review, the series stays op-only — the layer wrappers are dropped from the plan. Next up: `channels_first` support (#23267), then `output_size` auto-inference for `padding="valid"`

**Comment by hertschuh** [comment]
This was added, thanks @brianmanderson
